### PR TITLE
binding tests for volumes

### PR DESCRIPTION
**Pull request by baude**
add binding tests for volumes: inspect(get), create, remove, prune, and list

implement filters ability for volumes

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: libpod/volume.go [hunk 1]
```diff
@@ -126,3 +126,8 @@ func (v *Volume) GID() int {
 func (v *Volume) CreatedTime() time.Time {
 	return v.config.CreatedTime
 }
+
+// Config returns the volume's configuration.
+func (v *Volume) Config() *VolumeConfig {
+	return v.config
```
Comment (mheon) [inline libpod/volume.go]: use JSONDeepCopy to copy before returning

File: pkg/api/handlers/libpod/volumes.go [hunk 2]
```diff
@@ -76,25 +79,40 @@ func InspectVolume(w http.ResponseWriter, r *http.Request) {
...
+	if len(query.Filters) > 0 {
+		volumeFilters, err = generateVolumeFilters(query.Filters)
+		if err != nil {
+			utils.InternalServerError(w, err)
+			return
+		}
+	}
+	vols, err := runtime.Volumes(volumeFilters...)
+	if err != nil {
+		utils.InternalServerError(w, err)
```
Comment (mheon) [inline pkg/api/handlers/libpod/volumes.go]: Missing a return after here

File: pkg/api/handlers/types.go [hunk 3]
```diff
@@ -128,11 +128,16 @@ type CreateContainerConfig struct {
...
+// swagger:model VolumeCreate
 type VolumeCreateConfig struct {
-	Name   string            `json:"name"`
-	Driver string            `schema:"driver"`
-	Label  map[string]string `schema:"label"`
-	Opts   map[string]string `schema:"opts"`
+	// New volume's name. Can be left blank
+	Name string `json:"name"`
+	// Volume driver to use
+	Driver string `schema:"driver"`
```
Comment (mheon) [inline pkg/api/handlers/types.go]: FYI we don't support setting this in Libpod right now, will throw errors if it's passed. That's probably OK.

**Comment by baude** [comment]
fixed @mheon comments

File: pkg/api/handlers/libpod/volumes.go [hunk 4]
```diff
@@ -76,25 +84,46 @@ func InspectVolume(w http.ResponseWriter, r *http.Request) {
...
-	*/
+	var (
+		decoder       = r.Context().Value("decoder").(*schema.Decoder)
+		err           error
+		runtime       = r.Context().Value("runtime").(*libpod.Runtime)
+		volumeConfigs []*libpod.VolumeConfig
+		volumeFilters []libpod.VolumeFilter
+	)
+	query := struct {
+		Filters map[string][]string `json:"filter"`
```
Comment (jwhonce) [inline pkg/api/handlers/libpod/volumes.go]: `json:"filter"` should be `json:"filters"`

File: pkg/api/handlers/types.go [hunk 5]
```diff
@@ -128,11 +128,16 @@ type CreateContainerConfig struct {
...
 }
 
+// swagger:model VolumeCreate
 type VolumeCreateConfig struct {
-	Name   string            `json:"name"`
-	Driver string            `schema:"driver"`
-	Label  map[string]string `schema:"label"`
-	Opts   map[string]string `schema:"opts"`
+	// New volume's name. Can be left blank
+	Name string `json:"name"`
```
Comment (jwhonce) [inline pkg/api/handlers/types.go]: Shouldn't the tag be schema not json?

File: pkg/bindings/test/volumes_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		// Removing with a volume in use with force should work with a stopped container
+		zero := 0
+		err = containers.Stop(connText, "vtest", &zero)
+		Expect(err).To(BeNil())
+		err = volumes.Remove(connText, vol.Name, &trueFlag)
+		Expect(err).To(BeNil())
+	})
+
+	It("list volumes", func() {
+		// no volumes should be ok
```
Comment (jwhonce) [inline pkg/bindings/test/volumes_test.go]: Not sure why this test passed, given error in decoding above.

**Comment by baude** [comment]
fixed up

File: pkg/api/handlers/libpod/volumes.go [hunk 7]
```diff
@@ -76,25 +84,46 @@ func InspectVolume(w http.ResponseWriter, r *http.Request) {
...
-	*/
+	var (
+		decoder       = r.Context().Value("decoder").(*schema.Decoder)
+		err           error
+		runtime       = r.Context().Value("runtime").(*libpod.Runtime)
+		volumeConfigs []*libpod.VolumeConfig
+		volumeFilters []libpod.VolumeFilter
+	)
+	query := struct {
+		Filters map[string][]string `json:"filters"`
```
Comment (jwhonce) [inline pkg/api/handlers/libpod/volumes.go]: Tag should be `schema` not `json`

**Comment by baude** [comment]
fixed up

**Pull request merged**